mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs

```diff
@@ -1897,5 +1897,84 @@ internal static void SetLastWin32Error (int error)
 			throw new NotSupportedException();
 		}
 #endif
+
+		internal struct MarshalerInstanceKey {
+			public Type Type;
+			public string Cookie;
+
+			public override int GetHashCode () {
+				int result = 0;
+				if (Type != null)
+					result = Type.GetHashCode();
+				if (Cookie != null)
+					result ^= Cookie.GetHashCode();
+				return result;
+			}
+		}
+
+		internal class MarshalerInstanceKeyComparer : IEqualityComparer<MarshalerInstanceKey> {
+			public bool Equals (MarshalerInstanceKey lhs, MarshalerInstanceKey rhs) {
+				return (lhs.Type == rhs.Type) && (lhs.Cookie == rhs.Cookie);
+			}
+
+			public int GetHashCode (MarshalerInstanceKey key) {
+				return key.GetHashCode ();
+			}
+		}
+
+		internal static Dictionary<MarshalerInstanceKey, ICustomMarshaler> MarshalerInstanceCache = 
```

Comment:
This should be lazy initialized

Reply:
Is lazy initialization trivially thread safe? I guess I could guard it with a separate lock but I wouldn't be able to lazy initialize that. If there's somewhere else in the BCL we do this I can just copy that pattern

Reply:
Just use `LazyInitializer.EnsureInitialized`